### PR TITLE
Allow for stratification across categories, not considering each combination a stratum (see #2)

### PR DESCRIPTION
**Pull request by MartinBernstorff**
See #2 for context.

### CODE REVIEW

**Comment by MartinBernstorff** [comment]
@KennethEnevoldsen, unsure why the black action is failing. It says invalid `GITHUB_TOKEN`, but the action doesn't require it, nor is it in the repo secrets D:

I've requested a review from most of you because this code is pretty mission critical.

File: src/psycoptts/stratify_by_each_category_individually.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+
+def stratified_split_by_each_category(
+    df: DataFrame,
+    test_size=Union[float, int],
+    random_state=int,
+    stratify: Optional[List[str]] = None,
+) -> Tuple[DataFrame, DataFrame]:
+    no_cat_str = "no_cat"
+
```
Comment (HLasse) [inline src/psycoptts/stratify_by_each_category_individually.py:28]: add docstring

File: src/psycoptts/stratify_by_each_category_individually.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
+from collections import defaultdict
+import numpy as np
+from pandas import DataFrame
+from typing import Optional, List, Tuple, Union, Dict
+import random
+
+
+def stratified_split_by_each_category(
+    df: DataFrame,
+    test_size=Union[float, int],
```
Comment (HLasse) [inline src/psycoptts/stratify_by_each_category_individually.py]:
```suggestion
    test_size=float,
```

File: src/psycoptts/stratify_by_each_category_individually.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+) -> Tuple[DataFrame, DataFrame]:
+    no_cat_str = "no_cat"
+
+    # Generate target ns
+    target_n_in_test_by_cat = {
+        strat_col: df[df[strat_col] == 1].shape[0] * test_size for strat_col in stratify
+    }.copy()
+
+    if isinstance(test_size, float):
+        target_total_n_in_test = test_size * df.shape[0]
```
Comment (HLasse) [inline src/psycoptts/stratify_by_each_category_individually.py]:
```suggestion
    target_total_n_in_test = test_size * df.shape[0]
```

File: src/application/create_train_test_split.py [hunk 4]
```diff
@@ -56,27 +65,53 @@ def load_patient_ids(view="FOR_kohorte_demografi_inkl_2021_feb2022"):
             new_colname=outcome,
             id_colname="dw_ek_borger",
         )
+
+        n_with_outcome = combined_df[combined_df[outcome] == 1].shape[0]
+        unsplit_outcome_props[outcome] = round(n_with_outcome / n_in_split["total"], 4)
+
         msg.good(f"Added {outcome}")
 
     val_and_test_prop = 0.3
```
Comment (HLasse) [inline src/application/create_train_test_split.py]: call it `train_prop` instead

**Review by HLasse — CHANGES_REQUESTED** [review]
Good job! 🙌  Add a few more comments to the application script - otherwise good to go

File: src/application/create_train_test_split.py [hunk 5]
```diff
@@ -56,27 +65,53 @@ def load_patient_ids(view="FOR_kohorte_demografi_inkl_2021_feb2022"):
...
-    val_test_split = 0.5 
+    val_test_split = 0.5
     # Meaning that the prop of the dataset that ends in val is val_test_split * val_and_test_prop (e.g. 0.3 * 0.5 = 0.15)
 
-
-    X_train, X_intermediate = train_test_split(
+    msg.good("Starting train/intermediate split")
+    X_train, X_intermediate = stratified_split_by_each_category(
```
Comment (KennethEnevoldsen) [inline src/application/create_train_test_split.py:79]: This is fine, but wanted to mention that python typically use `_` to denote intermediate values. E.g. `X_` (not to be confused with just `_` which is a variables not intended to be used)

File: src/psycoptts/stratify_by_each_category_individually.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            )
+
+        idx = patient_ids_for_each_cat[cat_to_add_to].pop()
+
+        # If an idx belongs to multiple categories, also remove if from those that aren't currently cat_to_add_to
+        # This is __super__ inefficient (95% of current runtime), but running for 120.000 idxs takes 45 seconds.
+        # Acceptable runtime for current application.
+        remove_value_from_all_lists_it_exists_in(
+            value=idx, dict_of_lists=patient_ids_for_each_cat
+        )
```
Comment (KennethEnevoldsen) [inline src/psycoptts/stratify_by_each_category_individually.py]: An easy fix here is to have the list be sets (reduced the lookup complexity from O(n) to O(1)).

File: src/application/create_train_test_split.py [hunk 7]
```diff
@@ -56,27 +65,53 @@ def load_patient_ids(view="FOR_kohorte_demografi_inkl_2021_feb2022"):
...
         combined_df,
-        test_size=val_and_test_prop,
+        test_size=train_prop,
         random_state=random_state,
-        stratify=combined_df[outcomes],
+        stratify=outcomes,
     )
 
-    X_test, X_val = train_test_split(
+    msg.good("Starting test/val split")
```
Comment (KennethEnevoldsen) [inline src/application/create_train_test_split.py:86]:
```suggestion
    msg.info("Starting test/val split")
```

File: src/application/create_train_test_split.py [hunk 8]
```diff
@@ -56,27 +65,53 @@ def load_patient_ids(view="FOR_kohorte_demografi_inkl_2021_feb2022"):
...
+        test_outcome_prop = round(
+            X_test[X_test[outcome] == 1].shape[0] / n_in_split["test"], 4
+        )
+        val_outcome_prop = round(
+            X_val[X_val[outcome] == 1].shape[0] / n_in_split["val"], 4
+        )
+
+        print(
+            f"    (U|TEST|VAL|TRAIN): {unsplit_outcome_props[outcome]} | {test_outcome_prop} | {val_outcome_prop} | {train_outcome_prop} | {outcome}"
+        )
```
Comment (KennethEnevoldsen) [inline src/application/create_train_test_split.py]: brug msg

**Review by KennethEnevoldsen — APPROVED** [review]

**Pull request merged**